rust/main/chains/hyperlane-starknet/src/validator_announce.rs

```diff
@@ -0,0 +1,225 @@
+impl ValidatorAnnounce for StarknetValidatorAnnounce {
+    async fn get_announced_storage_locations(
+        &self,
+        validators: &[H256],
+    ) -> ChainResult<Vec<Vec<String>>> {
+        let validators_calldata: Vec<EthAddress> = validators
+            .iter()
+            .map(|v| {
+                TryInto::<FieldElement>::try_into(*v).map_err(Into::<HyperlaneStarknetError>::into)
+            })
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .map(EthAddress)
+            .collect();
+
+        let storage_locations_res = self
+            .contract
+            .get_announced_storage_locations(&validators_calldata)
+            .call()
+            .await
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+
+        let storage_locations = storage_locations_res
+            .into_iter()
+            .map(|validator_vec| {
+                validator_vec
+                    .into_iter()
+                    .map(|inner_vec| {
+                        inner_vec
+                            .into_iter()
+                            .map(|element| parse_cairo_short_string(&element))
+                            .collect::<Result<Vec<String>, ParseCairoShortStringError>>()
+                    })
+                    .collect::<Result<Vec<Vec<String>>, ParseCairoShortStringError>>()
+                    .map(|strings_vec| {
+                        strings_vec
+                            .into_iter()
+                            .map(|inner_vec| inner_vec.join(""))
+                            .collect::<Vec<String>>()
+                    })
+            })
+            .collect::<Result<Vec<Vec<String>>, ParseCairoShortStringError>>()
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+
+        Ok(storage_locations)
+    }
+
+    #[instrument(ret, skip(self))]
+    async fn announce_tokens_needed(&self, announcement: SignedType<Announcement>) -> Option<U256> {
+        let validator = bytes_to_hex(&announcement.value.validator.to_vec());
+
+        let Ok((_, max_cost)) = self.announce_contract_call(announcement).await else {
+            trace!("Unable to get announce contract call");
+            return None;
+        };
+
+        let Ok(balance) = self.provider.get_balance(validator).await else {
+            trace!("Unable to query balance");
+            return None;
+        };
+
+        let max_cost_u256: U256 = max_cost.into();
+
+        Some(max_cost_u256.saturating_sub(balance).into())
+    }
+
+    async fn announce(&self, announcement: SignedType<Announcement>) -> ChainResult<TxOutcome> {
+        let (contract_call, _) = self.announce_contract_call(announcement).await?;
+        let tx = contract_call
+            .send()
+            .await
+            .map_err(|e| HyperlaneStarknetError::AccountError(e.to_string()))?;
+        let invoke_tx_receipt =
+            get_transaction_receipt(&self.provider.rpc_client(), tx.transaction_hash).await;
```

Comment:
I'm seeing a pattern where txs are sent with `send` and then the receipt is fetched, maybe a good idea to refactor these into one `send_and_confirm` function or similar

Reply:
done